File: Python/models/note.py
from typing import Any
from datetime import datetime


class Note:
    def __init__(self, title: str, content: str, tags: list[str], created_on: datetime, last_updated_on: datetime) -> None:
        self._title = self._validate_type("title", title, str)
        self._content = self._validate_type("content", content, str)
        self._tags = self._validate_type("tags", tags, list)
        self._created_on = self._validate_type(
            "created_on", created_on, datetime)
        self._last_updated_on = self._validate_type(
            "last_updated_on", last_updated_on, datetime)

    @property
    def title(self) -> str:
        return self._title

    @property
    def content(self) -> str:
        return self._content

    @property
    def tags(self) -> list[str]:
        return self._tags

    @property
    def created_on(self) -> datetime:
        return self._created_on

    @property
    def last_updated_on(self) -> datetime:
        return self._last_updated_on

    @staticmethod
    def _validate_type(param_name: str, value: Any, expected_type: type) -> None:
        if not isinstance(value, expected_type):
            raise TypeError(
                f"{param_name} must be a {expected_type}, got {type(value).__name__}")
        return value

    def __str__(self) -> str:
        return f"Note(title={self.title}, content={self.content}, tags={self.tags}, created_on={self.created_on}, last_updated_on={self.last_updated_on})"

File: Go/models/note.go
package models

import (
	"encoding/json"
	"fmt"
	"time"
)

type Note struct {
	Title         string    `json:"title" bson:"title"`
	Content       string    `json:"content" bson:"content"`
	Tags          []string  `json:"tags" bson:"tags"`
	CreatedOn     time.Time `json:"created_on" bson:"created_on"`
	LastUpdatedOn time.Time `json:"last_updated_on" bson:"last_updated_on"`
}

func ParseDate(dateString string) (time.Time, error) {
	layouts := []string{
		"01/02/2006",
		"01/2/2006",
		"1/02/2006",
		"1/2/2006",
	}
	for _, layout := range layouts {
		if date, err := time.Parse(layout, dateString); err == nil {
			return date, nil
		}
	}
	return time.Time{}, fmt.Errorf("invalid date format: %s", dateString)
}

func (n *Note) UnmarshalJSON(data []byte) error {
	var aux struct {
		Title         string   `json:"title"`
		Content       string   `json:"content"`
		Tags          []string `json:"tags"`
		CreatedOn     string   `json:"created_on"`
		LastUpdatedOn string   `json:"last_updated_on"`
	}

	if err := json.Unmarshal(data, &aux); err != nil {
		return err
	}

	n.Title = aux.Title
	n.Content = aux.Content
	n.Tags = aux.Tags

	var err error
	n.CreatedOn, err = ParseDate(aux.CreatedOn)
	if err != nil {
		return err
	}
	n.LastUpdatedOn, err = ParseDate(aux.LastUpdatedOn)
	if err != nil {
		return err
	}

	return nil
}
